File: ts/src/jsonic.ts
/* Copyright (c) 2013-2026 Richard Rodger, MIT License */

/*  jsonic.ts
 *  Entry point and API.
 *
 *  jsonic = the `tabnas` parsing engine + the relaxed-JSON grammar +
 *  the BNF converter, wrapped in the historic `Jsonic` API: a parse
 *  function with the management methods attached as properties.
 *
 *  The lexer, parser, rule machinery, errors and utilities all live in
 *  the `tabnas` package now. This module no longer re-implements any of
 *  that — it just constructs `tabnas` engine instances and dresses them
 *  up in the callable-with-properties shape jsonic users expect.
 */

import {
  Tabnas,
  TabnasError as JsonicError,
  util,
  S,
  OPEN,
  CLOSE,
  BEFORE,
  AFTER,
  EMPTY,
  SKIP,
  makeLex,
  makeParser,
  makeToken,
  makePoint,
  makeRule,
  makeRuleSpec,
  makeFixedMatcher,
  makeSpaceMatcher,
  makeLineMatcher,
  makeStringMatcher,
  makeCommentMatcher,
  makeNumberMatcher,
  makeTextMatcher,
} from 'tabnas'

import { deep, assign, defprop, filterRules, parserwrap } from 'tabnas/utility'

import type {
  AltAction,
  AltCond,
  AltError,
  AltMatch,
  AltModifier,
  AltSpec,
  Bag,
  BnfConvertOptions,
  Config,
  Context,
  Counters,
  FuncRef,
  GrammarSetting,
  GrammarSpec,
  JsonicAPI,
  JsonicParse,
  Lex,
  LexCheck,
  LexMatcher,
  MakeLexMatcher,
  NormAltSpec,
  Options,
  Parser,
  Plugin,
  Point,
  Rule,
  RuleDefiner,
  RuleSpec,
  RuleSpecMap,
  RuleState,
  StateAction,
  Tin,
  Token,
} from './types'

import { defaults } from './defaults'

import { grammar, makeJSON } from './grammar'

import { bnf as bnfConvert } from './bnf'


// The full library type.
// NOTE: redeclared here (rather than imported) so the same name can be
// exported as both a type and the root instance value.
type Jsonic = JsonicParse & // A function that parses.
  JsonicAPI & { [prop: string]: any } // Plus the API methods. // Extensible by plugin decoration.


// Create a Jsonic instance.
//
//   make('jsonic')          -> the immutable root (restricted surface)
//   make('json')            -> the strict-JSON variant
//   make(options?, parent?) -> a fresh, customisable instance
//
// Each instance wraps a `tabnas` engine instance. The engine does the
// real work; this function just builds the callable + property surface
// and (re-)installs the relaxed-JSON grammar.
function make(param_options?: Bag | string, parent?: Jsonic): Jsonic {
  let injectFullAPI = true
  if ('jsonic' === param_options) {
    injectFullAPI = false
  } else if ('json' === param_options) {
    return makeJSON(root)
  }

  param_options = 'string' === typeof param_options ? {} : param_options

  // The underlying engine instance.
  let tabnas: Tabnas
  if (parent) {
    // A child engine: inherits the parent's merged options + config,
    // but starts with an empty rule set. The grammar and the parent's
    // plugins are re-applied below — against the *wrapper* — so that
    // plugin decorations land where the historic API expects them.
    const parentInst = parent.internal().tabnas as Tabnas
    tabnas = parentInst.make((param_options as Bag) || {})
  } else {
    const opts = deep(
      {},
      false === (param_options as Bag)?.defaults$ ? {} : defaults,
      (param_options as Bag) || {},
    )
    tabnas = new Tabnas(opts as any)
  }

  // Internal state record — mirrors the historic shape, but the live
  // values come straight off the engine instance (it swaps its parser
  // out on every options() change, hence the getters).
  const internal: any = {
    get parser() {
      return tabnas.internal().parser
    },
    get config() {
      return tabnas.internal().config
    },
    plugins: [] as Plugin[],
    get sub() {
      return tabnas.internal().sub
    },
    get mark() {
      return tabnas.internal().mark
    },
    get merged() {
      return tabnas.internal().merged
    },
    // Back-reference for child creation.
    tabnas,
  }

  // The primary parsing function. Drives the engine's parser directly
  // (rather than via tabnas.parse) so the *wrapper* — not the bare
  // engine instance — is what rule actions and plugins see as
  // `ctx.inst()`.
  const jsonic: any = function Jsonic(
    src: any,
    meta?: any,
    parent_ctx?: any,
  ): any {
    if (S.string === typeof src) {
      const opts_parser = (tabnas.options as any).parser
      const parser =
        opts_parser?.start ? parserwrap(opts_parser) : tabnas.internal().parser
      return parser.start(src, jsonic, meta, parent_ctx)
    }
    return src
  }

  // The API surface. Most members forward to the engine instance; for
  // methods that conventionally return "this", we return the wrapper so
  // chaining and `jsonic.use(p)('src')` keep working. `bnf` is jsonic's
  // own addition.
  const api: JsonicAPI = {
    parse: jsonic,

    token: tabnas.token as any,
    tokenSet: tabnas.tokenSet as any,
    fixed: tabnas.fixed as any,

    // Dual-shape callable: set via call, read via property. Sharing the
    // engine's own `options` object means option changes reconfigure
    // the engine (and swap its parser) as before.
    options: tabnas.options as any,

    config: () => tabnas.config(),

    // TODO: how to handle null plugin?
    use: function use(plugin: Plugin, plugin_options?: Bag): Jsonic {
      if (S.function !== typeof plugin) {
        throw new Error(
          'Jsonic.use: the first argument must be a function ' +
          'defining a plugin. See https://jsonic.senecajs.org/plugin',
        )
      }

      // Plugin name keys in options.plugin are the lower-cased plugin
      // function name.
      const plugin_name = plugin.name.toLowerCase()
      const full_plugin_options = deep(
        {},
        plugin.defaults || {},
        plugin_options || {},
      )

      ji.options({
        plugin: {
          [plugin_name]: full_plugin_options,
        },
      })
      const merged_plugin_options = ji.options.plugin[plugin_name]
      internal.plugins.push(plugin)
      plugin.options = merged_plugin_options

      return plugin(ji, merged_plugin_options) || ji
    },

    rule: (name?: string, define?: RuleDefiner | null) => {
      const r = tabnas.rule(name, define)
      return (r === tabnas ? jsonic : r) as any
    },

    make: (options?: Options | string) => make(options as any, jsonic),

    empty: (options?: Options) =>
      make({
        defaults$: false,
        standard$: false,
        grammar$: false,
        ...((options as Bag) || {}),
      }),

    id:
      'Jsonic/' +
      Date.now() +
      '/' +
      ('' + Math.random()).substring(2, 8).padEnd(6, '0') +
      (null == (tabnas.options as any).tag
        ? ''
        : '/' + (tabnas.options as any).tag),

    toString: () => api.id,

    sub: (spec: { lex?: any; rule?: any }) => {
      tabnas.sub(spec)
      return jsonic
    },

    util,

    grammar: (gs: GrammarSpec | string, setting?: GrammarSetting) => {
      if ('string' === typeof gs) {
        const parsed = make()(gs)
        if (null == parsed || 'object' !== typeof parsed) {
          return jsonic
        }
        gs = parsed as GrammarSpec
      }
      tabnas.grammar(gs as GrammarSpec, setting)
      return jsonic
    },

    // Convert a BNF grammar string into a jsonic GrammarSpec and install
    // it on this instance. Returns the generated spec so callers can
    // inspect, serialise or diff it. Use `bnf.toSpec(src, opts)` to
    // build the spec without installing it.
    bnf: (() => {
      const impl = (src: string, opts?: BnfConvertOptions) => {
        const spec = bnfConvert(src, opts)
        ;(ji as any).grammar(spec)
        return spec
      }
      impl.toSpec = (src: string, opts?: BnfConvertOptions) =>
        bnfConvert(src, opts)
      return impl
    })(),
  } as JsonicAPI

  // `api.make` reports its name as 'make' even though the enclosing
  // function is also named `make`.
  defprop(api.make, S.name, { value: S.make })

  // Assemble the public surface.
  if (injectFullAPI) {
    assign(jsonic, api)
  } else {
    // The immutable root exposes only a parse-and-fork surface
    // directly. NOTE: api.test.js pins this exact key set (plus the
    // deconstruction names added below).
    assign(jsonic, {
      empty: api.empty,
      parse: api.parse,
      sub: api.sub,
      id: api.id,
      toString: api.toString,
    })
  }

  // Hide internals where you can still find them.
  defprop(jsonic, 'internal', { value: () => internal })

  // The object the relaxed-JSON grammar (and re-run plugins) decorate.
  // For a full instance it is the instance itself; for the immutable
  // root it is a throwaway full-API view sharing the same engine.
  const ji: any = injectFullAPI
    ? jsonic
    : assign(Object.create(jsonic), api)

  if (parent) {
    // Transfer extra parent properties (preserves plugin decorations).
    for (const k in parent) {
      if (undefined === jsonic[k]) {
        jsonic[k] = (parent as any)[k]
      }
    }
    jsonic.parent = parent
  }

  // Install the relaxed-JSON grammar (unless suppressed).
  if (false !== (tabnas.options as any).grammar$) {
    grammar(ji)
  }

  // Re-run inherited plugins on the (now grammar-bearing) child, in
  // registration order, so option-conditional grammar alts get
  // re-evaluated against this instance's options.
  if (parent) {
    const inherited: Plugin[] = parent.internal().plugins
    for (const plugin of inherited) {
      ;(ji as any).use(plugin)
    }
  }

  // Apply rule.include / rule.exclude once the grammar and all plugins
  // have contributed their alts. The engine's RuleSpec adder no longer
  // filters in place (it returns a fresh spec instead), so the strict
  // 'json' variant — and any `make({ rule: { include/exclude } })` —
  // needs this final pass. Mirrors Tabnas#make's own filtering step.
  const cfg = tabnas.internal().config
  if (0 < cfg.rule.include.length || 0 < cfg.rule.exclude.length) {
    const tparser: any = tabnas.internal().parser
    const rsm = tparser.rule()
    const filtered: any = {}
    for (const rn of Object.keys(rsm)) {
      filtered[rn] = filterRules(rsm[rn], cfg)
    }
    tparser.rsm = filtered
    tparser.norm()
  }

  return jsonic
}


let root: any = undefined

// The global root Jsonic instance — its parsing rules cannot be
// modified. Use Jsonic.make() to create a modifiable instance.
let Jsonic: Jsonic = (root = make('jsonic'))

// Provide deconstruction export names.
root.Jsonic = root
root.JsonicError = JsonicError
root.makeLex = makeLex
root.makeParser = makeParser
root.makeToken = makeToken
root.makePoint = makePoint
root.makeRule = makeRule
root.makeRuleSpec = makeRuleSpec
root.makeFixedMatcher = makeFixedMatcher
root.makeSpaceMatcher = makeSpaceMatcher
root.makeLineMatcher = makeLineMatcher
root.makeStringMatcher = makeStringMatcher
root.makeCommentMatcher = makeCommentMatcher
root.makeNumberMatcher = makeNumberMatcher
root.makeTextMatcher = makeTextMatcher
root.OPEN = OPEN
root.CLOSE = CLOSE
root.BEFORE = BEFORE
root.AFTER = AFTER
root.EMPTY = EMPTY
root.SKIP = SKIP

root.util = util
root.make = make
root.S = S


// Export most of the engine types for use by plugins (re-exported from
// `tabnas` via ./types).
export type {
  AltAction,
  AltCond,
  AltError,
  AltMatch,
  AltModifier,
  AltSpec,
  Bag,
  BnfConvertOptions,
  Config,
  Context,
  Counters,
  FuncRef,
  GrammarSetting,
  GrammarSpec,
  Lex,
  LexCheck,
  LexMatcher,
  MakeLexMatcher,
  NormAltSpec,
  Options,
  Parser,
  Plugin,
  Point,
  Rule,
  RuleDefiner,
  RuleSpec,
  RuleSpecMap,
  RuleState,
  StateAction,
  Tin,
  Token,
}

export {
  // Jsonic is both a type and a value.
  Jsonic as Jsonic,
  JsonicError,
  util,
  make,
  makeToken,
  makePoint,
  makeRule,
  makeRuleSpec,
  makeLex,
  makeParser,
  makeFixedMatcher,
  makeSpaceMatcher,
  makeLineMatcher,
  makeStringMatcher,
  makeCommentMatcher,
  makeNumberMatcher,
  makeTextMatcher,
  OPEN,
  CLOSE,
  BEFORE,
  AFTER,
  EMPTY,
  SKIP,
  S,
  root,
}

export default Jsonic

if ('undefined' !== typeof module) {
  module.exports = Jsonic
}

File: go/jsonic.go
// Copyright (c) 2013-2026 Richard Rodger, MIT License

package tabnasjsonic

import (
	"fmt"
	"regexp"
	"sync"

	tjson "github.com/tabnas/json/go"
	tabnas "github.com/tabnas/parser/go"
)

// VERSION is this module's version. It MUST equal ts/package.json
// "version": the release orchestrator rewrites both, and
// TestVersionMatchesPackageJSON fails the build if they drift.
const VERSION = "0.6.4"

// grammarMark is the decoration key that records whether the relaxed-JSON
// grammar has already been installed on an instance. It guards the Grammar
// plugin so that re-running it (which the engine does on every SetOptions)
// does not rebuild the base rules and clobber caller rule modifications,
// while still letting Derive build the grammar fresh on a child (Derive
// copies decorations only after re-running plugins).
const grammarMark = "jsonic$grammar"

// jsonicOptions are the option overrides the grammar plugin layers on top
// of the engine's already-relaxed lexer defaults: the jsonic error
// identity, the `unprintable` string-error alignment matcher (see
// unprintable.go), and the comment markers held in *option* space.
// Error message templates and the rest of the relaxed lexer config are
// shared with the engine defaults.
//
// The comment defs mirror TS defaults.ts. The engine also lexes these
// markers by default, but only from config-space; declaring them as
// options (as TS does) makes them part of the option merge, so a caller
// adding a comment def extends this set instead of replacing it.
func jsonicOptions() Options {
	return Options{
		ErrMsg: &ErrMsgOptions{
			Name: "jsonic",
			Link: "https://github.com/tabnas/jsonic",
		},
		Lex: &LexOptions{
			Match: unprintableMatchSpecs(),
		},
		Comment: &CommentOptions{
			Def: commentDefDefaults(),
		},
	}
}

// commentDefDefaults are jsonic's default comment markers, mirroring the
// TS defaults.ts comment.def entries (hash / slash / multi).
func commentDefDefaults() map[string]*CommentDef {
	t, f := true, false
	return map[string]*CommentDef{
		"hash":  {Line: true, Start: "#", Lex: &t, EatLine: &f},
		"slash": {Line: true, Start: "//", Lex: &t, EatLine: &f},
		"multi": {Line: false, Start: "/*", End: "*/", Lex: &t, EatLine: &f},
	}
}

// normalizeCommentDefs aligns caller comment defs with the TS option
// merge: a partial def for a default name (hash / slash / multi)
// inherits the fields it leaves unset (start, end, line, lex, eatline)
// instead of replacing the whole definition — with Line false honored
// as an explicit block conversion when End is also set — a nil def
// stays nil (the removal marker), and a def for a new name is inactive
// unless it sets Lex — TS makeCommentMatcher reads `lex: !!om.lex`,
// while the Go engine would default an unset Lex to true. Needed
// because the engine's option Deep merge replaces map values wholesale
// per key.
func normalizeCommentDefs(o *Options) {
	if o == nil || o.Comment == nil || o.Comment.Def == nil {
		return
	}
	defs := commentDefDefaults()
	norm := make(map[string]*CommentDef, len(o.Comment.Def))
	for name, def := range o.Comment.Def {
		if def == nil {
			norm[name] = nil
			continue
		}
		if base, ok := defs[name]; ok {
			if merged, ok := Deep(base, def).(*CommentDef); ok {
				// Honor an explicit block conversion (TS `line: false`):
				// the struct merge treats the zero bool as unset and
				// would keep the default Line:true, but Line false with
				// an End marker is unambiguous — line comments never
				// use End.
				if !def.Line && def.End != "" {
					merged.Line = false
				}
				norm[name] = merged
				continue
			}
		}
		cp := *def
		if cp.Lex == nil {
			f := false
			cp.Lex = &f
		}
		norm[name] = &cp
	}
	cc := *o.Comment
	cc.Def = norm
	o.Comment = &cc
}

// Grammar is the idiomatic tabnas grammar plugin: it applies jsonic's
// option defaults and registers the relaxed-JSON grammar (the val / map /
// list / pair / elem rules) on the engine instance. Use it the standard
// way, and `use` it before any plugin that builds on jsonic's rules:
//
//	j := tabnas.Make()
//	j.Use(jsonic.Grammar)
//	j.Use(csv) // builds on jsonic's value/map/list rules
//	out, _ := j.Parse("a:1,b:[x,y,z]")
//
// The Jsonic-style helpers (Make, Parse, MakeJSON) are a legacy
// compatibility layer that installs this same plugin. To register the
// grammar rules without the jsonic branding (mirroring the TS
// `registerJsonicGrammar` export), use RegisterJsonicGrammar in grammar.go.
func Grammar(j *Jsonic, _ map[string]any) (err error) {
	// Never let grammar installation panic out of the plugin; report it.
	defer func() {
		if r := recover(); r != nil {
			err = fmt.Errorf("jsonic: Grammar plugin failed: %v", r)
		}
	}()
	j.SetOptions(jsonicOptions())
	return RegisterJsonicGrammar(j)
}

// grammarPlugin installs the relaxed-JSON rules without applying jsonic's
// option branding. It is what the legacy Make registers (Make applies the
// branding as a base option, before caller options, so a caller-supplied
// errmsg.name wins), and the body of the exported RegisterJsonicGrammar
// (grammar.go), the Go counterpart of TS `registerJsonicGrammar`. The
// decoration guard makes it idempotent under the engine's SetOptions plugin
// re-run while still letting Derive build the grammar fresh on a child.
func grammarPlugin(j *Jsonic, _ map[string]any) (err error) {
	// Convert any unexpected panic (e.g. from the @tabnas/json dependency)
	// into an error so plugin application never crashes the caller.
	defer func() {
		if r := recover(); r != nil {
			err = fmt.Errorf("jsonic: grammar installation failed: %v", r)
		}
	}()
	if mark, _ := j.Decoration(grammarMark).(bool); mark {
		return nil
	}
	j.Decorate(grammarMark, true)
	// Install the standard-JSON grammar core (val / map / list / pair /
	// elem) from the @tabnas/json plugin, then layer jsonic's relaxed
	// extensions on top of it.
	tjson.RegisterJSONGrammar(j)
	return buildGrammar(j.RSM(), j.Config())
}

// init registers jsonic as the engine's text parser so SetOptionsText and
// GrammarText (which parse a jsonic-format options/grammar string) work —
// the grammar-free engine has no parser of its own.
//
// The engine's text-options consumers (SetOptionsText, GrammarText,
// MapToOptions) expect a plain map[string]any tree and assert on it deeply.
// Since a jsonic parse now yields an insertion-ordered *OrderedMap object
// node, the registered parser flattens that node (recursively) back to a
// plain map before returning: an options/grammar spec is configuration, so
// its key order carries no meaning and dropping it keeps those consumers
// working unchanged.
func init() {
	tabnas.RegisterTextParser(func(src string) (any, error) {
		v, err := Make().Parse(src)
		if err != nil {
			return v, err
		}
		return plainNode(v), nil
	})
}

// plainNode recursively rewrites any *OrderedMap object node into a plain
// map[string]any (dropping insertion order) and copies []any elements,
// leaving all other values untouched. It is used only to feed the engine's
// map-shaped text-options consumers; jsonic's own Parse result keeps its
// *OrderedMap so callers still see source order.
func plainNode(v any) any {
	switch node := v.(type) {
	case *OrderedMap:
		m := make(map[string]any, len(node.Keys))
		for _, k := range node.Keys {
			m[k] = plainNode(node.Vals[k])
		}
		return m
	case map[string]any:
		m := make(map[string]any, len(node))
		for k, elem := range node {
			m[k] = plainNode(elem)
		}
		return m
	case []any:
		out := make([]any, len(node))
		for i, elem := range node {
			out[i] = plainNode(elem)
		}
		return out
	default:
		return v
	}
}

// Make creates a relaxed-JSON parser instance: a tabnas engine with the
// jsonic grammar plugin installed, plus any caller options on top.
// Equivalent to the historic jsonic.make().
func Make(opts ...Options) *Jsonic {
	var o Options
	if len(opts) > 0 {
		o = opts[0]
	}

	// Hold back rule include/exclude: they must filter the grammar only
	// after it is built, not at construction.
	base := o
	var inc, exc string
	if o.Rule != nil {
		inc, exc = o.Rule.Include, o.Rule.Exclude
		ruleCopy := *o.Rule
		ruleCopy.Include = ""
		ruleCopy.Exclude = ""
		base.Rule = &ruleCopy
	}

	// Merge partial comment defs with jsonic's defaults per name (the
	// whole-Options Deep below replaces map values wholesale per key).
	normalizeCommentDefs(&base)

	// Construct the engine with jsonic's branding as a base and caller
	// options merged on top (so a caller errmsg.name / tag / lexer setting
	// wins). Constructing with the merged options means the instance id,
	// config, and any option-conditional grammar alternates (list.child,
	// list.pair, …) are all decided against the final settings.
	merged, ok := Deep(jsonicOptions(), base).(Options)
	if !ok {
		// Deep merging two Options always yields Options; fall back rather
		// than panic on the impossible case.
		merged = base
	}
	j := tabnas.Make(merged)

	// Register (not just run) the grammar so the engine re-applies it when
	// deriving a child instance. The relaxed-JSON grammar is a plugin.
	// grammarPlugin never panics (it recovers internally); its error only
	// fires on an incompatible @tabnas/json core and cannot be surfaced
	// through Make's legacy *Jsonic signature — use the Grammar plugin
	// directly (it returns the error) when that matters.
	_ = j.Use(grammarPlugin, nil)

	if inc != "" || exc != "" {
		j.SetOptions(Options{Rule: &RuleOptions{Include: inc, Exclude: exc}})
	}
	return j
}

// Empty creates a parser instance with the jsonic configuration but no
// grammar rules, for building a grammar from scratch. Matches the historic
// jsonic.empty().
func Empty(opts ...Options) *Jsonic {
	j := Make(opts...)
	for _, rs := range j.RSM() {
		rs.Clear()
	}
	return j
}

// MakeJSON creates an instance configured to accept strict JSON only.
// Mirrors the TypeScript Jsonic.make('json'): it installs the full jsonic
// grammar, then restricts it to the json-tagged alternates and disables
// the lenient lexer features. Rejects unquoted keys/values, comments,
// hex/octal/binary numbers, trailing commas, leading-zero numbers,
// single-quoted or backtick strings, and empty input.
func MakeJSON() *Jsonic {
	f := false
	tr := true
	return Make(Options{
		Text: &TextOptions{Lex: &f},
		Number: &NumberOptions{
			Hex: &f, Oct: &f, Bin: &f,
			Sep:     "",
			Exclude: notStrictJSONNumber,
		},
		String: &StringOptions{
			Chars:        `"`,
			MultiChars:   "",
			AllowUnknown: &f,
			// No \xHH or \u{...}; strict JSON has only \uXXXX.
			EscapeStrict: &tr,
		},
		Comment: &CommentOptions{Lex: &f},
		Map:     &MapOptions{Extend: &f},
		Lex:     &LexOptions{Empty: &f},
		// Mirrors the TS `tokenSet: { KEY: ['#ST', null, null, null] }`:
		// strict JSON keys are quoted strings only, never #TX/#NR/#VL.
		// This is what makes `{1:1}` and `{null:null}` errors here, as
		// they are in TS and in encoding/json.
		TokenSet: map[string][]string{
			"KEY": {"#ST"},
		},
		Rule: &RuleOptions{
			Finish:  &f,
			Include: "json",
		},
	})
}

// strictJSONNumber is the RFC 8259 / ECMA-404 number grammar:
//
//	-? (0 | [1-9][0-9]*) (. [0-9]+)? ([eE] [-+]? [0-9]+)?
var strictJSONNumber = regexp.MustCompile(
	`^-?(0|[1-9][0-9]*)(\.[0-9]+)?([eE][-+]?[0-9]+)?$`)

// notStrictJSONNumber is the strict-JSON `number.exclude` predicate. The
// relaxed number matcher also accepts a leading `+`, leading zeros (`012`)
// and a bare leading or trailing `.` (`.123`, `1.`), none of which are
// valid JSON — exclude them so they fall through to the (disabled) text
// matcher and raise an error instead of parsing silently. Mirrors the TS
// `number.exclude` regexp in makeJSON (ts/src/grammar.ts).
func notStrictJSONNumber(s string) bool {
	return !strictJSONNumber.MatchString(s)
}

// defaultParser is a lazily-created instance reused by Parse, so repeated
// calls don't rebuild the engine and grammar each time. Building the jsonic
// grammar dominates a parse (see perf_test.go), so a rebuild-per-call Parse
// is many times slower than necessary. Parsing builds a fresh context per
// call and only reads instance state, so the shared instance is safe for
// concurrent use. Mirrors @tabnas/json's Parse.
var (
	defaultOnce   sync.Once
	defaultParser *Jsonic
)

// Parse parses a jsonic source string with default settings and returns
// the resulting value, or a *JsonicError on failure.
func Parse(src string) (any, error) {
	defaultOnce.Do(func() { defaultParser = Make() })
	return defaultParser.Parse(src)
}
